### ISSUE TITLE
Test with @Category on method not executable as "JUnit plug-in Test"

### ISSUE THREAD

**Issue by starfury-all** [junit-team/junit#1203]
Hi,

(non-parameterized) JUnit tests that have a `@Category` annotation on individual methods cannot be executed as **"JUnit plug-in Test"** if the test-suite is `@RunWith(Categories.class)`.
**[EDIT]** The problem also occurs when using "cascaded" test-suites and one of the *test-classes* is annotated, see [comment] (#issuecomment-142602805). **[/EDIT]**
Initialisation fails with:

    java.lang.Exception: Category annotations on Parameterized classes are not supported on individual methods.
      at org.junit.runners.model.InitializationError.<init>(InitializationError.java:38)
      at org.junit.experimental.categories.Categories.assertNoDescendantsHaveCategoryAnnotations(Categories.java:353)
      at org.junit.experimental.categories.Categories.assertNoDescendantsHaveCategoryAnnotations(Categories.java:355)
      at org.junit.experimental.categories.Categories.assertNoCategorizedDescendentsOfUncategorizeableParents(Categories.java:343)
      at org.junit.experimental.categories.Categories.<init>(Categories.java:318)
      ...

If the same (sample) testsuite is executed as normal "JUnit test" it is working fine.

## Environment ##
I came across this problem in Eclipse 4.3 (32 bit) with JUnit 4.11 but it's reproduceable it with the latest **Eclipse Mars 4.5.0 for RCP (64 bit)** with **JUnit 4.12**. 
All on *Windows 7.*

## Sample Classes ##
```java
public interface MyCategory { /* marker interface */ }
```
```java
@RunWith(Categories.class)
@SuiteClasses({MyTest.class})
public class MyTestsuite { /* test suite */ }
```
```java
public class MyTest {
    @Test
    @Category(MyCategory.class)
    public void testName() throws Exception { assertTrue(true); }
}
```

## What I found out so far ##
For me it looks like it's a `ClassLoader` problem:
When starting of the above test-suite the test initialization runs twice into 
[`Description.getTestClass() line: 288`](https://github.com/junit-team/junit/blob/master/src/main/java/org/junit/runner/Description.java#L288) for `MyTest`:
```java
fTestClass = Class.forName(name, false, getClass().getClassLoader());
```
When executing as normal *"JUnit Test"* the class is found and provided, but when executing as *"JUnit plug-in Test"* the class-loader cannot find the Class 'MyTest' and `null` is returned.
Probably due to Eclipse RCP Bundle boundaries.

Due to `Description.getTestClass()` returning `null` the method `Categories.canHaveCategorizedChildren(Description)` then decides that methods may not be annotated:

    Description.getTestClass() line: 288    
    Categories.canHaveCategorizedChildren(Description) line: 362    
    Categories.assertNoCategorizedDescendentsOfUncategorizeableParents(Description) line: 342    
    Categories.<init>(Class<?>, RunnerBuilder) line: 318    
    NativeConstructorAccessorImpl.newInstance0(Constructor, Object[]) line: not available [native method]    
    NativeConstructorAccessorImpl.newInstance(Object[]) line: 57    
    DelegatingConstructorAccessorImpl.newInstance(Object[]) line: 45    
    Constructor<T>.newInstance(Object...) line: 526    
    AnnotatedBuilder.buildRunner(Class<Runner>, Class<?>) line: 107    
    AnnotatedBuilder.runnerForClass(Class<?>) line: 86    
    AnnotatedBuilder(RunnerBuilder).safeRunnerForClass(Class<?>) line: 59    
    AllDefaultPossibilitiesBuilder.runnerForClass(Class<?>) line: 26    
    AllDefaultPossibilitiesBuilder(RunnerBuilder).safeRunnerForClass(Class<?>) line: 59    
    ClassRequest.getRunner() line: 33    
    JUnit4TestLoader.createUnfilteredTest(Class<?>, String[]) line: 84

It then throws the above mentioned ´InitializationError´.

## One more thing ##
It would be helpful if the method `Categories.assertNoDescendantsHaveCategoryAnnotations(Description)` would provide at least a little information about the problematic test-class (from the `Description` object at hand) when throwing the ´InitializationError´.
And maybe the misleading(?) reference to *"Paramterized classes"* should be removed.

**Comment by starfury-all** [comment]
I just noticed, that the current state of the class [Categories](https://github.com/junit-team/junit/blob/master/src/main/java/org/junit/experimental/categories/Categories.java) already lacks the methods `assertNoCategorizedDescendentsOfUncategorizeableParents` and `canHaveCategorizedChildren`, so the problem might already be 'fixed' ...?

**Comment by marcphilipp** [comment]
Can you check in your environment using the lastest [4.13-SNAPSHOT](https://oss.sonatype.org/content/repositories/snapshots/junit/junit/4.13-SNAPSHOT/junit-4.13-20150810.150727-40.jar)?

**Comment by starfury-all** [comment]
Before trying 4.13-SNAPSHOT, I just noticed another constellation where this bug occurs and would like to document it for completeness: Similar to the constellation " Suite / TestClass / (annotated) TestMethod " the bug also occurs in a constellation " (outer) Suite / (inner) Suite / (annotated) TestClass ".
In this constellation the `@Category` annotation on the class causes the `InitializationError` to be thrown.

## Sample Classes ##
```java
@RunWith(Categories.class)
@SuiteClasses({ TheInnerTestsuite.class })
public class TheOuterTestsuite { /* test suite */ }
```
```java
@RunWith(Categories.class)
@SuiteClasses({ MyTest.class })
public class TheInnerTestsuite { /* test suite */ }
```
```java
@Category(MyCategory.class)
public class TheAnnotatedTest {
	@Test
	public void testName() throws Exception { assertTrue(true); }
}
```

And now I'll go and try if this thing is fixed in 4.13.

**Comment by starfury-all** [comment]
Hi Marc,
the issue is fixed in the version you've linked above. Great!
Thanks.

**Comment by marcphilipp** [comment]
That's good to hear! Thanks for your detailed bug report and trying the latest snapshot!

**Comment by sethiraj** [comment]
Hi @marcphilipp,

This issue does not seem to be resolved in the latest 4.13-SNAPSHOT, as I am getting this error again.
Could you please let me know, when this could be resolved?

Regards,
Surendran Ethiraj

**Comment by marcphilipp** [comment]
@sethiraj Which snapshot JAR are you using?
